spec/requests/api_controller_authentication_spec.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ApiController, type: :request do
+  include_context 'with supplier with access token'
+  include_context 'with undefine mock controller'
+
+  let(:response_json) { JSON.parse(response.body) }
+  let(:schema) { load_yaml_schema('error_responses.yaml') }
+  let(:detail_401) { 'Token expired or invalid' }
+
+  let(:create_mock_controller) do
+    class MockController < ApiController
+      def custom
+        render json: { hello: 'world' }, status: :ok
+      end
+    end
+    Rails.application.routes.draw do
+      get '/custom', to: 'mock#custom'
+    end
+  end
+
+  around do |example|
+    create_mock_controller
+    get '/custom', headers: headers
+    example.run
+    undefine_mock_controller # NB: it is necessary to undefine the mock controller
```

Comment:
Rather than doing this, why not add `MockController` class to the code base - including the idempotentable module? The routing to that can be restricted to the specs only with:

```ruby
  namespace :api do
    constraints(->(_req) { Rails.env.test? }) do
      get 'custom', to: 'mock#custom`
    end
  end
```

This lets us add further test endpoints to the mock controller if we need to in future. I think I prefer that to defining code on the the fly like this (and then having to undefine it after each spec). Probably worth getting other opinions on this though.

Reply:
Hi @smoothcontract  I'm a bit wary about the suggested approach: it doesn't seem correct to add controllers to the codebase which are only there to facilitate testing: I think test code should belong with the specs, not with the application. I'll see if I can make the it work with a more concrete MockController defined outside of the app...

Reply:
It does smell a bit affecting the production runtime for testing controllers. I don't have super strong opinions, though. As long as it is isolated and potentially namespaced, I'm pretty confident we won't run into any issues even with eager loading, etc, in production. I think on balance I prefer the existing approach that clearly says what's getting defined for the test. I don't think this discussion should block the PR from flying, either.